contributors/design-proposals/scheduler-equivalence-class.md

```diff
@@ -0,0 +1,325 @@
+
+    - No `MatchInterPodAffinity`: the scheduler will make sure newly binded pod will not break the existing inter pod affinity. So we does not need to invalidate MatchInterPodAffinity when pod added. But when a pod is deleted, existing inter pod affinity may become invalid. (e.g. this pod was preferred by some else, or vice versa).
+
+        - NOTE: assumptions above **will not** stand when we implemented features like `RequiredDuringSchedulingRequiredDuringExecution`.
+
+    - No `NoDiskConflict`: the newly scheduled pod fits to existing pods on this node, it will also fits to equivalence class of existing pods.
+
+- **Scope:** 
+    - The node which the pod was binded with.
+
+
+
+- **Operation:**
+    - UPDATE
+
+- **Invalid predicates:**
+
+    - Only if `pod.NodeName` did not change (otherwise it drops to add/delete case)
+
+    - `GeneralPredicates` if the pod's resource requests are updated.
+
+    - `MatchInterPodAffinity` if the pod's labels are updated.
+
+- **Scope:**
+    - The node which the pod was binded with
+
+
+
+- **Operation:**
+    - DELETE
+
+- **Invalid predicates:**
+    - `MatchInterPodAffinity` if the pod's labels are updated.
+
+- **Scope:**
+    - All nodes in the same failure domain
+
+- **Invalid predicates:**
+
+    - `NoDiskConflict` if the pod has special volume like `RBD`, `ISCSI`, `GCEPersistentDisk` etc.
+
+- **Scope:**
+    - The node which the pod was binded with.
+
+
+### 3.5 Node
+
+
+- **Operation:**
+    - UPDATE
+
+- **Invalid predicates:**
+
+    - `GeneralPredicates`, if `node.Status.Allocatable` or node labels changed.
+
+    - `ServiceAffinity`, if node labels changed, since selector result may change.
+
+    - `MatchInterPodAffinity`, if value of label changed, since any node label can be topology key of pod.
+
+    - `NoVolumeZoneConflict`, if zone related label change.
+
+    - `PodToleratesNodeTaints`, if node taints changed.
+
+    - `CheckNodeMemoryPressure`, `CheckNodeDiskPressure`, `CheckNodeCondition`, if related node condition changed.
+
+- **Scope:**
+    - The updated node.
+
+- **Operation:**
+    - DELETE
+
+- **Invalid predicates:**
+    - All predicates
+
+- **Scope:**
+    - The deleted node
+
+
+# Notes for scheduler developers
+
+1. When implementing a new predicate, developers are expected to check how the result of your new predicate function can be affected by related API object change (add/delete/update), and consider to invalidate this predicate for related API object change in scheduler/factory/factory.go.
+
+2. When updating a existing predicate, developers should consider whether your change will introduce new dependency to attributes of some API objects like Pod, Node, Service etc. If yes, developer should consider to invalidate this predicate for related API object change in `scheduler/factory/factory.go`.
```

Comment:
Please consider:

When updating an existing predicate, developers should consider whether their changes introduce new dependency on attributes of any API objects like Pod, Node, Service, etc. If so, developer should consider invalidating caches results of this predicate in `scheduler/factory/factory.go`.